### PR TITLE
support to specify devices that will be accessible within the service container

### PR DESCRIPTION
**Pull request by wlwgcdxc**
support for features like this `docker run --device=/dev/snd:/dev/snd:mrw ... `

### CODE REVIEW

File: master/engine/docker/engine.go [hunk 1]
```diff
@@ -144,3 +150,23 @@ func (e *dockerEngine) Run(cfg openedge.ServiceInfo, vs map[string]openedge.Volu
...
+func (e *dockerEngine) parseDeviceSpecs(devices []string) (deviceBindings []container.DeviceMapping, err error) {
+	for _, rawDevice := range devices {
+		deviceMapping := strings.Split(rawDevice, ":")
+		device := container.DeviceMapping{}
+		switch len(deviceMapping) {
+		case 3:
+			device.CgroupPermissions = deviceMapping[2]
+		case 2:
+			device.CgroupPermissions = "mrw"
+		default:
```
Comment (wangxiaochen04) [inline master/engine/docker/engine.go]: 好像case 1的情况下，PathOnHost = PathInContainer

File: master/engine/docker/engine.go [hunk 2]
```diff
@@ -144,3 +150,23 @@ func (e *dockerEngine) Run(cfg openedge.ServiceInfo, vs map[string]openedge.Volu
 	}
 	return s, nil
 }
+
+func (e *dockerEngine) parseDeviceSpecs(devices []string) (deviceBindings []container.DeviceMapping, err error) {
+	for _, rawDevice := range devices {
+		deviceMapping := strings.Split(rawDevice, ":")
+		device := container.DeviceMapping{}
+		switch len(deviceMapping) {
```
Comment (ludanfeng) [inline master/engine/docker/engine.go]: deviceParts := strings.Split(device, ":")
deviceMapping := container.DeviceMapping{}

File: master/engine/docker/engine.go [hunk 3]
```diff
@@ -144,3 +150,23 @@ func (e *dockerEngine) Run(cfg openedge.ServiceInfo, vs map[string]openedge.Volu
...
+		case 3:
+			device.CgroupPermissions = deviceMapping[2]
+		case 2:
+			device.CgroupPermissions = "mrw"
+		default:
+			err = fmt.Errorf("error device mapping[%s]", rawDevice)
```
Comment (ludanfeng) [inline master/engine/docker/engine.go]: err = fmt.Errorf("invalid  device mapping (%s)", device)

File: master/engine/docker/engine.go [hunk 4]
```diff
@@ -144,3 +150,23 @@ func (e *dockerEngine) Run(cfg openedge.ServiceInfo, vs map[string]openedge.Volu
 	}
 	return s, nil
 }
+
+func (e *dockerEngine) parseDeviceSpecs(devices []string) (deviceBindings []container.DeviceMapping, err error) {
```
Comment (ludanfeng) [inline master/engine/docker/engine.go]: 可以移除 (e *dockerEngine)

**Pull request merged**